File: Offer/CPP/3.cpp
/**
*  struct ListNode {
*        int val;
*        struct ListNode *next;
*        ListNode(int x) :
*              val(x), next(NULL) {
*        }
*  };
*/

// 借助栈来实现
class Solution
{
  public:
    vector<int> printListFromTailToHead(ListNode *head)
    {
        vector<int> result;
        stack<int> stk;
        ListNode *p = head;
        while (p != nullptr)
        {
            stk.push(p -> val);
            p = p -> next;
        }
        while (!stk.empty())
        {
            result.push_back(stk.top());
            stk.pop();
        }
        return result;
    }
};

// 直接使用 vector 的 insert 方法，每次在开始处插入元素
class Solution
{
  public:
    vector<int> printListFromTailToHead(ListNode *head)
    {
        vector<int> array;
        if (head != nullptr)
        {
            // 在 array 开始处插入元素
            array.insert(array.begin(), head -> val);
            // 从下一个开始
            while (head -> next != nullptr)
            {
                array.insert(array.begin(), head -> next -> val);
                head = head -> next;
            }
        }
        return array;
    }
};

File: Offer/Python/3.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


# 列表可以理解为栈
class Solution:
    # 返回从尾部到头部的列表值序列，例如[1,2,3]
    def printListFromTailToHead(self, listNode):
        res = []
        while listNode:
            res.append(listNode.val)
            listNode = listNode.next
        return res[::-1]


# 使用列表的插入操作
class Solution:
    # 返回从尾部到头部的列表值序列，例如[1,2,3]
    def printListFromTailToHead(self, listNode):
        res = []
        while listNode:
            res.insert(0, listNode.val)
            listNode = listNode.next
        return res
